File: js/gem-geometry.js
/**
 * gem-geometry.js
 *
 * Generates a 57-facet round brilliant-cut diamond as a THREE.BufferGeometry.
 * Each facet is a separate flat-shaded polygon (non-indexed, per-face normals)
 * so that individual facets catch and reflect light independently — essential
 * for realistic diamond rendering.
 *
 * Facet layout (standard round brilliant):
 *   Crown (top):  1 table + 8 star + 8 bezel (kite) + 16 upper girdle = 33
 *   Pavilion (bottom):  8 main + 16 lower girdle = 24
 *   Total: 57 facets
 *
 * Coordinate system: Y-up. Table faces +Y, culet points -Y.
 * Girdle sits at Y = 0 (crown above, pavilion below).
 */

import * as THREE from "three";

const DEG2RAD = Math.PI / 180;

/**
 * Load the parametric config from JSON and build the geometry.
 * @param {string} jsonPath - URL to brilliant-cut.json
 * @returns {Promise<THREE.BufferGeometry>}
 */
export async function createBrilliantGeometry(jsonPath) {
    const resp = await fetch(jsonPath);
    const config = await resp.json();
    return buildGeometry(config.parameters);
}

/**
 * Build the geometry directly from parameter values (no fetch).
 * @param {object} params
 * @returns {THREE.BufferGeometry}
 */
export function buildGeometry(params) {
    const {
        girdleRadius = 1.0,
        tableRatio = 0.57,
        crownAngleDeg = 34.5,
        pavilionAngleDeg = 40.75,
        girdleThickness = 0.025,
        symmetryFold = 8,
        starFacetLengthRatio = 0.50,
        lowerGirdleFacetLengthRatio = 0.75,
        culetSizeRatio = 0.0,
    } = params;

    const N = symmetryFold; // 8-fold rotational symmetry
    const R = girdleRadius;
    const tableR = R * tableRatio;

    // Derived heights
    const crownAngle = crownAngleDeg * DEG2RAD;
    const pavilionAngle = pavilionAngleDeg * DEG2RAD;
    const crownHeight = (R - tableR) * Math.tan(crownAngle);
    const pavilionDepth = R * Math.tan(pavilionAngle);
    const halfGirdle = girdleThickness / 2;

    // Y coordinates
    const yTable = crownHeight + halfGirdle;
    const yGirdleTop = halfGirdle;
    const yGirdleBot = -halfGirdle;
    const yCulet = -(pavilionDepth + halfGirdle);

    // ---------------------------------------------------------------
    // Helper: point on circle at angle theta, radius r, at height y
    // ---------------------------------------------------------------
    function pt(r, theta, y) {
        return new THREE.Vector3(r * Math.cos(theta), y, r * Math.sin(theta));
    }

    // ---------------------------------------------------------------
    // Key angular positions (2*N evenly spaced around the gem)
    // Main facet vertices sit at i * (2PI / 2N) = i * PI/N
    // ---------------------------------------------------------------
    const step = (2 * Math.PI) / (2 * N); // PI/8 for N=8

    // Girdle vertices: 2N points on the girdle circle (top and bottom edge)
    const girdleTop = [];
    const girdleBot = [];
    for (let i = 0; i < 2 * N; i++) {
        const angle = i * step;
        girdleTop.push(pt(R, angle, yGirdleTop));
        girdleBot.push(pt(R, angle, yGirdleBot));
    }

    // Table vertices: N points on the table circle
    // Table vertices align with even-indexed girdle vertices (main directions)
    const tableVerts = [];
    for (let i = 0; i < N; i++) {
        const angle = i * 2 * step; // same as i * (2PI/N)
        tableVerts.push(pt(tableR, angle, yTable));
    }

    // Star facet intermediate points on the crown (between table and girdle)
    // These sit at odd multiples of step (halfway between main girdle vertices)
    // at a radius interpolated between table and girdle, at a height on the
    // crown slope.
    const starR = tableR + (R - tableR) * starFacetLengthRatio;
    const starY = yTable - (starR - tableR) * Math.tan(crownAngle);
    const starVerts = [];
    for (let i = 0; i < N; i++) {
        const angle = (2 * i + 1) * step;
        starVerts.push(pt(starR, angle, starY));
    }

    // Pavilion main facet lower points (toward culet)
    // For culetSizeRatio = 0 they all converge to the culet point.
    const culetR = R * culetSizeRatio;
    const culetVerts = [];
    if (culetR < 0.001) {
        // Single culet point
        culetVerts.push(new THREE.Vector3(0, yCulet, 0));
    } else {
        for (let i = 0; i < N; i++) {
            const angle = i * 2 * step;
            culetVerts.push(pt(culetR, angle, yCulet));
        }
    }

    // Lower girdle intermediate points on the pavilion
    // Sit at odd multiples of step, at a radius interpolated between girdle and culet
    const lowerR = R * (1 - lowerGirdleFacetLengthRatio);
    const lowerY = yGirdleBot - (R - lowerR) * Math.tan(pavilionAngle);
    const lowerVerts = [];
    for (let i = 0; i < N; i++) {
        const angle = (2 * i + 1) * step;
        lowerVerts.push(pt(lowerR, angle, lowerY));
    }

    // ---------------------------------------------------------------
    // Collect all triangles (non-indexed, for flat per-face normals)
    // ---------------------------------------------------------------
    const triangles = [];

    function addTri(a, b, c) {
        triangles.push(a.clone(), b.clone(), c.clone());
    }

    function addQuad(a, b, c, d) {
        // Two triangles: (a, b, c) and (a, c, d)
        addTri(a, b, c);
        addTri(a, c, d);
    }

    // ---------- TABLE (1 octagonal facet → 8 triangles) ----------
    // Table center
    const tableCenter = new THREE.Vector3(0, yTable, 0);
    for (let i = 0; i < N; i++) {
        const next = (i + 1) % N;
        addTri(tableCenter, tableVerts[i], tableVerts[next]);
    }

    // ---------- STAR FACETS (8 triangles) ----------
    // Each star facet is a triangle: tableVerts[i], starVerts[i], tableVerts[next]
    for (let i = 0; i < N; i++) {
        const next = (i + 1) % N;
        addTri(tableVerts[next], tableVerts[i], starVerts[i]);
    }

    // ---------- BEZEL / KITE FACETS (8 quadrilaterals → 16 triangles) ----------
    // Each bezel kite connects: starVerts[prev], girdleTop[2*i], starVerts[i], tableVerts[i]
    // Kite shape: top = tableVerts[i], left = starVerts[prev], bottom = girdleTop[2*i], right = starVerts[i]
    for (let i = 0; i < N; i++) {
        const prev = (i - 1 + N) % N;
        addQuad(
            tableVerts[i],
            starVerts[prev],
            girdleTop[2 * i],
            starVerts[i]
        );
    }

    // ---------- UPPER GIRDLE FACETS (16 triangles) ----------
    // Two triangles per main direction sector:
    //   Left:  starVerts[prev], girdleTop[2*i - 1], girdleTop[2*i]
    //   Right: starVerts[i],    girdleTop[2*i],     girdleTop[2*i + 1]
    for (let i = 0; i < N; i++) {
        const gi = 2 * i;
        const giPrev = (gi - 1 + 2 * N) % (2 * N);
        const giNext = (gi + 1) % (2 * N);
        const prev = (i - 1 + N) % N;

        addTri(starVerts[prev], girdleTop[giPrev], girdleTop[gi]);
        addTri(starVerts[i], girdleTop[gi], girdleTop[giNext]);
    }

    // ---------- GIRDLE (thin band — 2N quads → 4N triangles) ----------
    // Connect girdleTop[i] to girdleBot[i]
    for (let i = 0; i < 2 * N; i++) {
        const next = (i + 1) % (2 * N);
        addQuad(
            girdleTop[i],
            girdleBot[i],
            girdleBot[next],
            girdleTop[next]
        );
    }

    // ---------- PAVILION MAIN FACETS (8 quadrilaterals → 16 triangles) ----------
    // Each main pavilion facet is a kite from girdleBot[2*i] down to culet,
    // flanked by lowerVerts.
    // For culetSizeRatio = 0, culet is a single point → these become triangles.
    if (culetR < 0.001) {
        const culet = culetVerts[0];
        for (let i = 0; i < N; i++) {
            const prev = (i - 1 + N) % N;
            addQuad(
                girdleBot[2 * i],
                lowerVerts[prev],
                culet,
                lowerVerts[i]
            );
        }
    } else {
        for (let i = 0; i < N; i++) {
            const prev = (i - 1 + N) % N;
            addQuad(
                girdleBot[2 * i],
                lowerVerts[prev],
                culetVerts[i],
                lowerVerts[i]
            );
        }
    }

    // ---------- LOWER GIRDLE FACETS (16 triangles) ----------
    for (let i = 0; i < N; i++) {
        const gi = 2 * i;
        const giPrev = (gi - 1 + 2 * N) % (2 * N);
        const giNext = (gi + 1) % (2 * N);
        const prev = (i - 1 + N) % N;

        addTri(lowerVerts[prev], girdleBot[giPrev], girdleBot[gi]);
        addTri(lowerVerts[i], girdleBot[gi], girdleBot[giNext]);
    }

    // ---------------------------------------------------------------
    // Build BufferGeometry
    // ---------------------------------------------------------------
    const positions = new Float32Array(triangles.length * 3);
    for (let i = 0; i < triangles.length; i++) {
        positions[i * 3] = triangles[i].x;
        positions[i * 3 + 1] = triangles[i].y;
        positions[i * 3 + 2] = triangles[i].z;
    }

    const geometry = new THREE.BufferGeometry();
    geometry.setAttribute("position", new THREE.BufferAttribute(positions, 3));
    geometry.computeVertexNormals();

    // Override with flat per-face normals:
    // For each triangle, compute the face normal and assign it to all 3 vertices.
    const normals = geometry.attributes.normal.array;
    const vA = new THREE.Vector3();
    const vB = new THREE.Vector3();
    const vC = new THREE.Vector3();
    const ab = new THREE.Vector3();
    const ac = new THREE.Vector3();
    const faceNormal = new THREE.Vector3();

    for (let i = 0; i < triangles.length; i += 3) {
        vA.set(positions[i * 3], positions[i * 3 + 1], positions[i * 3 + 2]);
        vB.set(positions[(i + 1) * 3], positions[(i + 1) * 3 + 1], positions[(i + 1) * 3 + 2]);
        vC.set(positions[(i + 2) * 3], positions[(i + 2) * 3 + 1], positions[(i + 2) * 3 + 2]);

        ab.subVectors(vB, vA);
        ac.subVectors(vC, vA);
        faceNormal.crossVectors(ab, ac).normalize();

        for (let j = 0; j < 3; j++) {
            normals[(i + j) * 3] = faceNormal.x;
            normals[(i + j) * 3 + 1] = faceNormal.y;
            normals[(i + j) * 3 + 2] = faceNormal.z;
        }
    }

    geometry.attributes.normal.needsUpdate = true;
    geometry.computeBoundingSphere();

    return geometry;
}

File: vulkan/src/scene/gem_geometry.c
/**
 * gem_geometry.c
 *
 * Port of the JavaScript 57-facet brilliant-cut geometry builder.
 *
 * Facet layout (standard round brilliant):
 *   Crown (top):  1 table (8 tri) + 8 star + 8 bezel kite (16 tri) + 16 upper girdle = 33 facets
 *   Girdle:       16 quads (32 tri band)
 *   Pavilion:     8 main (16 tri) + 16 lower girdle = 24 facets
 *   Total: 57 facets, 112 triangles, 336 vertices
 *
 * Coordinate system: Y-up. Table faces +Y, culet points -Y.
 * Girdle sits at Y = 0.
 */

#include "gem_geometry.h"
#include "util/file_io.h"

#include <cjson/cJSON.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define DEG2RAD (3.14159265358979323846f / 180.0f)
#define PI      3.14159265358979323846f

// Max triangles: table(8) + star(8) + bezel(16) + upper_girdle(16)
//              + girdle(32) + pavilion_main(16) + lower_girdle(16) = 112
#define MAX_TRIANGLES 112
#define MAX_VERTICES  (MAX_TRIANGLES * 3)

// ---------------------------------------------------------------------------
// Internal: 3D point helper
// ---------------------------------------------------------------------------
typedef struct Vec3 {
    float x, y, z;
} Vec3;

static Vec3 vec3_make(float x, float y, float z)
{
    return (Vec3){ x, y, z };
}

// Point on circle at angle theta, radius r, height y
static Vec3 pt(float r, float theta, float y)
{
    return vec3_make(r * cosf(theta), y, r * sinf(theta));
}

static Vec3 vec3_sub(Vec3 a, Vec3 b)
{
    return vec3_make(a.x - b.x, a.y - b.y, a.z - b.z);
}

static Vec3 vec3_cross(Vec3 a, Vec3 b)
{
    return vec3_make(
        a.y * b.z - a.z * b.y,
        a.z * b.x - a.x * b.z,
        a.x * b.y - a.y * b.x
    );
}

static Vec3 vec3_normalize(Vec3 v)
{
    float len = sqrtf(v.x * v.x + v.y * v.y + v.z * v.z);
    if (len > 1e-8f) {
        float inv = 1.0f / len;
        v.x *= inv;
        v.y *= inv;
        v.z *= inv;
    }
    return v;
}

static Vec3 vec3_scale(Vec3 v, float s)
{
    return vec3_make(v.x * s, v.y * s, v.z * s);
}

// ---------------------------------------------------------------------------
// Triangle accumulator
// ---------------------------------------------------------------------------
typedef struct TriAccum {
    Vec3     verts[MAX_VERTICES];
    uint32_t count;  // number of vertices (always multiple of 3)
} TriAccum;

static void accum_tri(TriAccum *acc, Vec3 a, Vec3 b, Vec3 c)
{
    if (acc->count + 3 > MAX_VERTICES) return;
    acc->verts[acc->count++] = a;
    acc->verts[acc->count++] = b;
    acc->verts[acc->count++] = c;
}

static void accum_quad(TriAccum *acc, Vec3 a, Vec3 b, Vec3 c, Vec3 d)
{
    accum_tri(acc, a, b, c);
    accum_tri(acc, a, c, d);
}

// ---------------------------------------------------------------------------
// Default parameters
// ---------------------------------------------------------------------------

GemCutParams gem_cut_params_default(void)
{
    return (GemCutParams){
        .girdleRadius              = 1.0f,
        .tableRatio                = 0.57f,
        .crownAngleDeg             = 34.5f,
        .pavilionAngleDeg          = 40.75f,
        .girdleThickness           = 0.025f,
        .symmetryFold              = 8,
        .starFacetLengthRatio      = 0.50f,
        .lowerGirdleFacetLengthRatio = 0.75f,
        .culetSizeRatio            = 0.0f,
    };
}

// ---------------------------------------------------------------------------
// JSON loading
// ---------------------------------------------------------------------------

GemCutParams gem_cut_params_from_json(const char *path)
{
    GemCutParams defaults = gem_cut_params_default();

    char *json = file_io_read_text(path);
    if (!json) {
        fprintf(stderr, "[gem_geometry] Failed to read %s, using defaults.\n", path);
        return defaults;
    }

    cJSON *root = cJSON_Parse(json);
    free(json);

    if (!root) {
        fprintf(stderr, "[gem_geometry] JSON parse error, using defaults.\n");
        return defaults;
    }

    cJSON *params = cJSON_GetObjectItem(root, "parameters");
    if (!params) {
        cJSON_Delete(root);
        return defaults;
    }

    GemCutParams p = defaults;

    cJSON *item;
    if ((item = cJSON_GetObjectItem(params, "girdleRadius")))
        p.girdleRadius = (float)item->valuedouble;
    if ((item = cJSON_GetObjectItem(params, "tableRatio")))
        p.tableRatio = (float)item->valuedouble;
    if ((item = cJSON_GetObjectItem(params, "crownAngleDeg")))
        p.crownAngleDeg = (float)item->valuedouble;
    if ((item = cJSON_GetObjectItem(params, "pavilionAngleDeg")))
        p.pavilionAngleDeg = (float)item->valuedouble;
    if ((item = cJSON_GetObjectItem(params, "girdleThickness")))
        p.girdleThickness = (float)item->valuedouble;
    if ((item = cJSON_GetObjectItem(params, "symmetryFold")))
        p.symmetryFold = item->valueint;
    if ((item = cJSON_GetObjectItem(params, "starFacetLengthRatio")))
        p.starFacetLengthRatio = (float)item->valuedouble;
    if ((item = cJSON_GetObjectItem(params, "lowerGirdleFacetLengthRatio")))
        p.lowerGirdleFacetLengthRatio = (float)item->valuedouble;
    if ((item = cJSON_GetObjectItem(params, "culetSizeRatio")))
        p.culetSizeRatio = (float)item->valuedouble;

    cJSON_Delete(root);
    return p;
}

// ---------------------------------------------------------------------------
// Geometry builder
// ---------------------------------------------------------------------------

bool gem_geometry_build(const GemCutParams *params, float scale,
                         GemGeometry *out)
{
    memset(out, 0, sizeof(*out));

    const int   N            = params->symmetryFold;  // 8
    const float R            = params->girdleRadius;
    const float tableR       = R * params->tableRatio;
    const float crownAngle   = params->crownAngleDeg * DEG2RAD;
    const float pavilionAngle = params->pavilionAngleDeg * DEG2RAD;
    const float halfGirdle   = params->girdleThickness / 2.0f;

    // Derived heights
    const float crownHeight  = (R - tableR) * tanf(crownAngle);
    const float pavilionDepth = R * tanf(pavilionAngle);

    // Y coordinates
    const float yTable     = crownHeight + halfGirdle;
    const float yGirdleTop = halfGirdle;
    const float yGirdleBot = -halfGirdle;
    const float yCulet     = -(pavilionDepth + halfGirdle);

    // Angular step: 2*N points around the girdle
    const float step = (2.0f * PI) / (2.0f * (float)N);  // PI / N

    // --- Girdle vertices (2N top + 2N bottom) ---
    Vec3 girdleTop[32], girdleBot[32];  // max 2*16
    for (int i = 0; i < 2 * N; i++) {
        float angle = (float)i * step;
        girdleTop[i] = pt(R, angle, yGirdleTop);
        girdleBot[i] = pt(R, angle, yGirdleBot);
    }

    // --- Table vertices (N points) ---
    Vec3 tableVerts[16];
    for (int i = 0; i < N; i++) {
        float angle = (float)i * 2.0f * step;
        tableVerts[i] = pt(tableR, angle, yTable);
    }

    // --- Star vertices (N points, at odd multiples of step) ---
    const float starR = tableR + (R - tableR) * params->starFacetLengthRatio;
    const float starY = yTable - (starR - tableR) * tanf(crownAngle);
    Vec3 starVerts[16];
    for (int i = 0; i < N; i++) {
        float angle = (float)(2 * i + 1) * step;
        starVerts[i] = pt(starR, angle, starY);
    }

    // --- Culet vertices ---
    const float culetR = R * params->culetSizeRatio;
    Vec3 culetVerts[16];
    int culetCount = 0;
    if (culetR < 0.001f) {
        culetVerts[0] = vec3_make(0.0f, yCulet, 0.0f);
        culetCount = 1;
    } else {
        for (int i = 0; i < N; i++) {
            float angle = (float)i * 2.0f * step;
            culetVerts[i] = pt(culetR, angle, yCulet);
        }
        culetCount = N;
    }

    // --- Lower girdle intermediate vertices ---
    const float lowerR = R * (1.0f - params->lowerGirdleFacetLengthRatio);
    const float lowerY = yGirdleBot - (R - lowerR) * tanf(pavilionAngle);
    Vec3 lowerVerts[16];
    for (int i = 0; i < N; i++) {
        float angle = (float)(2 * i + 1) * step;
        lowerVerts[i] = pt(lowerR, angle, lowerY);
    }

    // --- Accumulate triangles ---
    TriAccum acc = { .count = 0 };

    // TABLE (1 octagonal facet -> N triangles)
    Vec3 tableCenter = vec3_make(0.0f, yTable, 0.0f);
    for (int i = 0; i < N; i++) {
        int next = (i + 1) % N;
        accum_tri(&acc, tableCenter, tableVerts[i], tableVerts[next]);
    }

    // STAR FACETS (N triangles)
    for (int i = 0; i < N; i++) {
        int next = (i + 1) % N;
        accum_tri(&acc, tableVerts[next], tableVerts[i], starVerts[i]);
    }

    // BEZEL / KITE FACETS (N quads -> 2N triangles)
    for (int i = 0; i < N; i++) {
        int prev = (i - 1 + N) % N;
        accum_quad(&acc,
                   tableVerts[i],
                   starVerts[prev],
                   girdleTop[2 * i],
                   starVerts[i]);
    }

    // UPPER GIRDLE FACETS (2N triangles)
    for (int i = 0; i < N; i++) {
        int gi     = 2 * i;
        int giPrev = (gi - 1 + 2 * N) % (2 * N);
        int giNext = (gi + 1) % (2 * N);
        int prev   = (i - 1 + N) % N;

        accum_tri(&acc, starVerts[prev], girdleTop[giPrev], girdleTop[gi]);
        accum_tri(&acc, starVerts[i], girdleTop[gi], girdleTop[giNext]);
    }

    // GIRDLE BAND (2N quads -> 4N triangles)
    for (int i = 0; i < 2 * N; i++) {
        int next = (i + 1) % (2 * N);
        accum_quad(&acc,
                   girdleTop[i],
                   girdleBot[i],
                   girdleBot[next],
                   girdleTop[next]);
    }

    // PAVILION MAIN FACETS (N quads -> 2N triangles)
    if (culetR < 0.001f) {
        Vec3 culet = culetVerts[0];
        for (int i = 0; i < N; i++) {
            int prev = (i - 1 + N) % N;
            accum_quad(&acc,
                       girdleBot[2 * i],
                       lowerVerts[prev],
                       culet,
                       lowerVerts[i]);
        }
    } else {
        for (int i = 0; i < N; i++) {
            int prev = (i - 1 + N) % N;
            accum_quad(&acc,
                       girdleBot[2 * i],
                       lowerVerts[prev],
                       culetVerts[i],
                       lowerVerts[i]);
        }
    }

    // LOWER GIRDLE FACETS (2N triangles)
    for (int i = 0; i < N; i++) {
        int gi     = 2 * i;
        int giPrev = (gi - 1 + 2 * N) % (2 * N);
        int giNext = (gi + 1) % (2 * N);
        int prev   = (i - 1 + N) % N;

        accum_tri(&acc, lowerVerts[prev], girdleBot[giPrev], girdleBot[gi]);
        accum_tri(&acc, lowerVerts[i], girdleBot[gi], girdleBot[giNext]);
    }

    // --- Build vertex array with flat per-face normals ---
    uint32_t triCount = acc.count / 3;
    out->vertexCount = acc.count;
    out->vertices = malloc(out->vertexCount * sizeof(GemVertex));
    if (!out->vertices) {
        fprintf(stderr, "[gem_geometry] Failed to allocate %u vertices.\n", out->vertexCount);
        return false;
    }

    for (uint32_t t = 0; t < triCount; t++) {
        Vec3 a = acc.verts[t * 3 + 0];
        Vec3 b = acc.verts[t * 3 + 1];
        Vec3 c = acc.verts[t * 3 + 2];

        // Apply uniform scale
        a = vec3_scale(a, scale);
        b = vec3_scale(b, scale);
        c = vec3_scale(c, scale);

        // Flat face normal from winding order
        Vec3 ab = vec3_sub(b, a);
        Vec3 ac = vec3_sub(c, a);
        Vec3 n  = vec3_normalize(vec3_cross(ab, ac));

        // Outward-normal correction: the normal must point away from the gem
        // center (origin).  Use the face centroid as the reference direction.
        // If n points inward (dot < 0), reverse the winding (swap b and c)
        // so Vulkan's CCW front-face rule sees the correct orientation, and
        // negate n to match.
        Vec3 centroid = vec3_make(
            (a.x + b.x + c.x) / 3.0f,
            (a.y + b.y + c.y) / 3.0f,
            (a.z + b.z + c.z) / 3.0f
        );
        float outwardness = n.x * centroid.x + n.y * centroid.y + n.z * centroid.z;
        if (outwardness < 0.0f) {
            // Swap b and c to flip CCW winding, then negate normal
            Vec3 tmp = b;
            b = c;
            c = tmp;
            n.x = -n.x;
            n.y = -n.y;
            n.z = -n.z;
        }

        for (int v = 0; v < 3; v++) {
            Vec3 *p;
            if (v == 0) p = &a;
            else if (v == 1) p = &b;
            else p = &c;

            GemVertex *vert = &out->vertices[t * 3 + v];
            vert->position[0] = p->x;
            vert->position[1] = p->y;
            vert->position[2] = p->z;
            vert->normal[0]   = n.x;
            vert->normal[1]   = n.y;
            vert->normal[2]   = n.z;
        }
    }

    fprintf(stderr, "[gem_geometry] Built %u triangles (%u vertices)\n",
            triCount, out->vertexCount);

    return true;
}

void gem_geometry_destroy(GemGeometry *geo)
{
    free(geo->vertices);
    memset(geo, 0, sizeof(*geo));
}
